FirebaseRemoteConfig/SwiftNew/RemoteConfigComponent.swift

```diff
@@ -0,0 +1,160 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+import FirebaseCore
+import FirebaseCoreExtension
+import FirebaseRemoteConfigInterop
+
+// TODO(ncooke3): Once Obj-C tests are ported, all `public` access modifers can be removed.
+
+// TODO(ncooke3): Move to another pod.
+@objc protocol FIRAnalyticsInterop {}
+
+/// Provides and creates instances of Remote Config based on the namespace provided. Used in the
+/// interop registration process to keep track of RC instances for each `FIRApp` instance.
+@objc(FIRRemoteConfigProvider) public protocol RemoteConfigProvider {
+  /// Cached instances of Remote Config objects.
+  var instances: [String: RemoteConfig] { get set }
+
+  /// Default method for retrieving a Remote Config instance, or creating one if it doesn't exist.
+  func remoteConfig(forNamespace remoteConfigNamespace: String) -> RemoteConfig?
+}
+
+/// A concrete implementation for FIRRemoteConfigInterop to create Remote Config instances and
+/// register with Core's component system.
+@objc(FIRRemoteConfigComponent) public final class RemoteConfigComponent: NSObject {
+  // Because Component now need to register two protocols (provider and interop), we need a way to
+  // return the same component instance for both registered protocol, this singleton pattern allow
+  // us
+  // to return the same component object for both registration callback.
+  static var componentInstances: [String: RemoteConfigComponent] = [:]
+  static let componentInstancesLock = NSLock()
+
+  /// The FIRApp that instances will be set up with.
+  @objc public weak var app: FirebaseApp?
+
+  /// Cached instances of Remote Config objects.
+  public var instances: [String: RemoteConfig]
+  let instancesLock = NSLock()
+
+  /// Default initializer.
+  @objc public init(app: FirebaseApp) {
+    self.app = app
+    instances = [:]
+    super.init()
+  }
+}
+
+extension RemoteConfigComponent: RemoteConfigProvider {
+  /// Default method for retrieving a Remote Config instance, or creating one if it doesn't exist.
+  @objc public func remoteConfig(forNamespace remoteConfigNamespace: String) -> RemoteConfig? {
+    guard let app else {
+      return nil
+    }
+
+    let errorPropertyName = if app.options.googleAppID.isEmpty {
+      "googleAppID"
+    } else if app.options.gcmSenderID.isEmpty {
+      "GCMSenderID"
+    } else if (app.options.projectID ?? "").isEmpty {
+      "projectID"
+    } else { nil as String? }
+
+    if let errorPropertyName {
+      NSException.raise(
```

Comment:
Hmm, we've been converting NSExceptions to fatal errors, but this approach might be better?

Reply:
Good catch. I didn't find much but I suspect that `NSException.raise` may be recoverable in ObjC but not in Swift.

Reply:
True, but I'm not sure we want to support recovering from these.

Reply:
Agreed. Switching to fatalError.

Reply:
This actually becomes problematic since the ObjC assert that the API throws an exception. Swift's fatal error cannot be caught. I'm going to revert back to the NSException with a TODO to re-evaluate the exception behavior